src/test/scala/io/iohk/ethereum/blockchain/sync/CacheBasedBlacklistSpec.scala

```diff
@@ -0,0 +1,84 @@
+package io.iohk.ethereum.blockchain.sync
+
+import akka.actor.ActorSystem
+import akka.testkit.TestKit
+import io.iohk.ethereum.WithActorSystemShutDown
+import io.iohk.ethereum.network.PeerId
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.wordspec.AnyWordSpecLike
+
+import scala.concurrent.duration._
+import com.google.common.testing.FakeTicker
+import com.github.blemale.scaffeine.Scaffeine
+import java.util.concurrent.TimeUnit
+
+class CacheBasedBlacklistSpec extends AnyWordSpecLike with Matchers {
+  import Blacklist._
+
+  private val peer1 = PeerId("1")
+  private val peer2 = PeerId("2")
+  private val peer3 = PeerId("3")
+  private val peer4 = PeerId("4")
+  private val peer5 = PeerId("5")
+
+  private val reason = BlacklistReason.ErrorInBlockHeaders
+  private val anotherReason = BlacklistReason.BlockBodiesNotMatchingHeaders
+
+  private def withBlacklist(maxElements: Int)(test: CacheBasedBlacklist => Unit): Unit = {
+    val blacklist = CacheBasedBlacklist.empty(maxElements)
+    test(blacklist)
+  }
+
+  "CacheBasedBlacklist" should {
+    "add elements and respect max number of elements" in withBlacklist(3) { blacklist =>
+      blacklist.add(peer1, 1.minute, reason)
+      blacklist.add(peer2, 1.minute, reason)
+      blacklist.add(peer3, 1.minute, anotherReason)
+      blacklist.add(peer4, 1.minute, anotherReason)
+      blacklist.add(peer5, 1.minute, reason)
+      blacklist.cache.cleanUp()
+      assert(blacklist.keys.size <= 3)
```

Comment:
Any reason for a single assertion that will pass even if `blacklist.keys.size` is zero?

Reply:
If you look at the eviction description of caffeine you'll see that it's not guaranteed to be exactly `maxElements` (so 3 in this case).

Reply:
...but I added a `> 0` :+1: